### ISSUE TITLE
animate() - callback function is executed before animation has been completed in latest jquery 3.2.1

### ISSUE THREAD

**Issue by keerthanaRajendran** [jquery/jquery#3861]
Hi All,

We have upgraded jquery version from 3.1.1 to 3.2.1. Now the code inside animation()- callback function is executed  before animation is completed and then it executes the next line 

$(".btn2").click(function(){
        $("p").slideDown(1, function(){
        console.log("inside")
        });
        console.log("outside")
    });
JQuery 3.2.1 output
      1.inside
      2.outside.

JQuery 3.1.1 output
      1.outside
      2.inside. 

Sample Link(jquery 3.2.1): https://www.w3schools.com/code/tryit.asp?filename=FLMCAWA173W6

Sample Link(jquery 3.1.1): https://www.w3schools.com/code/tryit.asp?filename=FLMETVQCUR42

Expected result: 
1.outside.
2.Inside

Please let me know if you have any concern,

Thanks for your help in advance.

Regards,
Keerthana.

**Comment by gibson042** [comment]
This appears to be a case of code changes affecting the most likely outcome of a race condition. jQuery animations can resolve synchronously, and a 1 millisecond duration is short enough that it has almost always elapsed by the time we check for completion in the first tick as of https://github.com/jquery/jquery/commit/6d43dc42337089f5fb52b715981c12993f490920 (on my computer, 3 ms seems to always avoid the synchronous completion and 2 ms is pretty much a coin flip). However, I'm not convinced that this is a problem with jQuery.

If you _require_ asynchrony, I recommend using `setTimeout` or `.then` ([example](https://www.w3schools.com/code/tryit.asp?filename=FLVYLBLI4XSY)) inside the completion callback. But you should also reconsider having such a short animation in the first place—it would probably be cleaner to set the properties immediately _without_ animation and construct your own Promise for asynchronous handling.

**Comment by gibson042** [comment]
We decided in the meeting to leave this open. Hopefully, we can abandon synchronous resolution by updating [`jQuery.fx.start`](https://github.com/jquery/jquery/blob/731c501155ef139f53029c0e58409b80f0af3a0c/src/effects.js#L684) to schedule the first tick with a zero-duration timeout, which will also fix the stutter from animating a _collection_ (in which [the first element gets a head start](https://jsfiddle.net/2v6cyorf/4/)).

**Comment by hdsantos89** [comment]
Yeah I have been having this issue for a number of hours. I thought I was using the callback function incorrectly.

The callback function is being called before the animation is complete. Because I need to check the element's scrollLeft() position value after the animation is finished, I resorted to using the callback function. Any workarounds?
